src/charm.py

```diff
@@ -5,49 +5,422 @@
         else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
+            self.container.stop("exporter")
+        self.unit.status = ops.ActiveStatus()
+
+    def _gen_pebble_plan(self) -> ops.pebble.LayerDict:
+        """Generate penpot pebble plan.
+
+        Returns:
+            Penpot pebble plan.
+        """
+        plan = ops.pebble.LayerDict(
+            summary="penpot services",
+            description="penpot services",
+            services={
+                "frontend": {
+                    "command": './nginx-entrypoint.sh nginx -g "daemon off;"',
+                    "working-dir": "/opt/penpot/frontend/",
+                    "override": "replace",
+                    "after": ["backend"],
+                    "environment": {
+                        "PENPOT_BACKEND_URI": "http://127.0.0.1:6060",
+                        "PENPOT_EXPORTER_URI": self._get_penpot_exporter_uri(),
+                        "PENPOT_INTERNAL_RESOLVER": self._get_local_resolver(),
+                        "PENPOT_FLAGS": " ".join(self._get_penpot_frontend_options()),
+                    },
+                },
+                "backend": {
+                    "command": "/opt/penpot/backend/run.sh",
+                    "override": "replace",
+                    "working-dir": "/opt/penpot/backend/",
+                    "environment": {
+                        "JAVA_HOME": "/usr/lib/jvm/java-21-openjdk-amd64",
+                        "PENPOT_TELEMETRY_ENABLED": "false",
+                        "PENPOT_PUBLIC_URI": typing.cast(str, self._get_public_uri()),
+                        "PENPOT_FLAGS": " ".join(self._get_penpot_backend_options()),
+                        **typing.cast(dict[str, str], self._get_penpot_secret_key()),
+                        **typing.cast(dict[str, str], self._get_postgresql_credentials()),
+                        **typing.cast(dict[str, str], self._get_redis_credentials()),
+                        **typing.cast(dict[str, str], self._get_smtp_credentials()),
+                        **typing.cast(dict[str, str], self._get_s3_credentials()),
+                    },
+                },
+                "exporter": {
+                    "command": "node app.js",
+                    "working-dir": "/opt/penpot/exporter/",
+                    "override": "replace",
+                    "after": ["backend", "frontend"],
+                    "environment": {
+                        "PENPOT_PUBLIC_URI": "http://127.0.0.1:8080",
+                        "PLAYWRIGHT_BROWSERS_PATH": "/opt/penpot/exporter/browsers",
+                        **typing.cast(dict[str, str], self._get_redis_credentials()),
+                    },
+                },
+            },
+        )
+        return plan
+
+    def _check_ready(self) -> bool:  # pylint: disable=too-many-return-statements
```

Comment:
We could avoid this by adding messages/problems to a list, and if that list is non-empty, combining all the items into the BlockedStatus message and returning `False` once. e.g "Waiting for redis, s3"

Reply:
Good idea, added, thanks!